### PR TITLE
fix(types): output plugin typings

### LINKED ISSUE #3365: Typings for ProtractorPlugin missing from index.d.ts in Protractor 4.0.0
Hi there,

It seems like the definition of the `ProtractorPlugin` is no longer accessible to plugin developers.

This definition used to be available from the `protractor/built/plugins` package as per the [docs here](https://github.com/angular/protractor/blob/master/docs/plugins.md#writing-plugins-in-typescript):

```typescript
import {ProtractorPlugin} from 'protractor/built/plugins';
```

but as Protractor 4.0.0 does not include this definition in the main `built/index.d.ts` nor does it provide definitions at `protractor/built/plugins/`, the class is beyond reach.

Could you please expose it in the main definitions file please?

Many thanks,
Jan

### PR DESCRIPTION
**Pull request by cnishina**
closes #3365

### CODE REVIEW

**Comment by cnishina** [comment]
change protractor plugins to an interface; however, more direction is needed for:

```
addSuccess: ({specName: 'hello world'})
```

**Comment by cnishina** [comment]
@sjelin will need input for above ^^

File: docs/plugins.md [hunk 1]
```diff
@@ -312,17 +312,26 @@ export function onPageLoad(): void {
...
     this.addSuccess({specName: 'Hello, World!'});
-  };
+  }
 };
+
+module.exports = myPlugin;
+
 ```
 
 Then, in a separate file, export an instance of that class using an assignment
```
Comment (sjelin) [inline docs/plugins.md]: Delete the part about the separate file now yeah?

File: lib/plugins.ts [hunk 2]
```diff
@@ -18,16 +18,217 @@ export interface PluginConfig {
...
+   * The plugin configuration object. Note that this is not the entire
+   * Protractor config object, just the entry in the plugins array for this
+   * plugin.
+   *
+   * @type {Object}
+   */
+  config?: PluginConfig;
+
+  /**
+   * Adds a failed assertion to the test's results.
```
Comment (sjelin) [inline lib/plugins.ts]: Mention that this is added by the protractor API - not to be implemented by the plugin author

File: lib/plugins.ts [hunk 3]
```diff
@@ -18,16 +18,217 @@ export interface PluginConfig {
...
+   * @throws {Error} Throws an error if called after results have been reported
+   */
+  addFailure?:
       (message?: string,
        info?: {specName?: string, stackTrace?: string}) => void;
-  addSuccess: (info?: {specName?: string}) => void;
-  addWarning: (message?: string, info?: {specName?: string}) => void;
+
+  /**
+   * Adds a passed assertion to the test's results.
```
Comment (sjelin) [inline lib/plugins.ts]: Mention that this is added by the protractor API - not to be implemented by the plugin author

File: lib/plugins.ts [hunk 4]
```diff
@@ -18,16 +18,217 @@ export interface PluginConfig {
...
+   *       - specName The name of the spec which this assertion belongs to.
+   *            Defaults to `PLUGIN_NAME + ' Plugin Tests'`.
+   *     Defaults to `{}`.
+   *
+   * @throws {Error} Throws an error if called after results have been reported
+   */
+  addSuccess?: (info?: {specName?: string}) => void;
+
+  /**
+   * Warns the user that something is problematic.
```
Comment (sjelin) [inline lib/plugins.ts]: Mention that this is added by the protractor API - not to be implemented by the plugin author

**Comment by sjelin** [comment]
Should be good with some minor doc fixes

**Comment by sjelin** [comment]
LGTM

**Pull request merged**